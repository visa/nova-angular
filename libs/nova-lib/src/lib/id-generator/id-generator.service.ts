/**
 *              Copyright (c) 2025 Visa, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 **/
/** This was inspired by materials id-generator: https://github.com/angular/components/blob/main/src/cdk/a11y/id-generator.ts */
import { APP_ID, inject, Injectable } from '@angular/core';

/**
 * Keeps track of the ID count per prefix. This helps us make the IDs a bit more deterministic.
 */
const counters: Record<string, number> = {};

/**
 * Service that generates unique IDs for DOM nodes.
 * Note that this service should not be used inside templates and should only be used within classes.
 * @docs {@link https://design.visa.com/angular/services/id-generator | See docs}
 */
@Injectable({ providedIn: 'root' })
export class IdGenerator {
  private appId: null | string = inject(APP_ID, { optional: true });

  /**
   * Generates a unique ID with a specific prefix.
   * @param prefix Prefix to add to the ID.
   */
  public newId(prefix: string = ''): string {
    // Omit the app ID if it's the default `ng`. Since the vast majority of pages have one
    // Angular app on them, we can reduce the amount of breakages by not adding it.
    if (this.appId && this.appId !== 'ng') prefix = `${this.appId}-${prefix}`;
    if (!counters.hasOwnProperty(prefix)) counters[prefix] = 0;
    return `${prefix}-${counters[prefix]++}`;
  }
}
