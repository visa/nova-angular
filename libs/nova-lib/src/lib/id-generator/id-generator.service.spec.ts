/**
 *              Copyright (c) 2025 Visa, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 **/
import { APP_ID, Component, inject } from '@angular/core';
import { TestBed } from '@angular/core/testing';
import { render } from '@testing-library/angular';

import { IdGenerator } from './id-generator.service';

@Component({
  imports: [],
  providers: [IdGenerator],
  selector: 'test-component',
  standalone: true,
  template: `<div [id]="id1"></div>
    <div [id]="id2"></div>`
})
class TestComponent {
  readonly idGenerator = inject(IdGenerator);

  readonly id1 = this.idGenerator.newId();
  readonly id2 = this.idGenerator.newId();
}

describe('IdGenerator', () => {
  it('should be created', () => {
    TestBed.configureTestingModule({
      providers: [IdGenerator, { provide: APP_ID, useValue: 'testApp' }]
    });
    const service = TestBed.inject(IdGenerator);
    expect(service).toBeTruthy();
  });

  it('should generate unique IDs with a specific prefix', () => {
    TestBed.configureTestingModule({
      providers: [IdGenerator, { provide: APP_ID, useValue: 'testApp' }]
    });
    const service = TestBed.inject(IdGenerator);
    const prefix = 'testPrefix';
    const id1 = service.newId(prefix);
    const id2 = service.newId(prefix);
    expect(id1).toBe('testApp-testPrefix-0');
    expect(id2).toBe('testApp-testPrefix-1');
  });

  it('should generate unique IDs without a specific prefix', () => {
    TestBed.configureTestingModule({
      providers: [IdGenerator, { provide: APP_ID, useValue: 'testApp' }]
    });
    const service = TestBed.inject(IdGenerator);
    const id1 = service.newId();
    const id2 = service.newId();
    expect(id1).toBe('testApp--0');
    expect(id2).toBe('testApp--1');
  });

  it('should omit the app ID if it is "ng"', () => {
    TestBed.configureTestingModule({
      providers: [IdGenerator, { provide: APP_ID, useValue: 'ng' }]
    });
    const service = TestBed.inject(IdGenerator);
    const prefix = 'testPrefix';
    const id1 = service.newId(prefix);
    const id2 = service.newId(prefix);
    expect(id1).toBe('testPrefix-0');
    expect(id2).toBe('testPrefix-1');
  });

  it('should generate IDs independently for different prefixes', () => {
    TestBed.configureTestingModule({
      providers: [IdGenerator, { provide: APP_ID, useValue: 'testApp' }]
    });
    const service = TestBed.inject(IdGenerator);
    const prefix1 = 'prefix1';
    const prefix2 = 'prefix2';
    const id1 = service.newId(prefix1);
    const id2 = service.newId(prefix2);
    const id3 = service.newId(prefix1);
    const id4 = service.newId(prefix2);
    expect(id1).toBe('testApp-prefix1-0');
    expect(id2).toBe('testApp-prefix2-0');
    expect(id3).toBe('testApp-prefix1-1');
    expect(id4).toBe('testApp-prefix2-1');
  });

  it('should handle empty prefix', () => {
    TestBed.configureTestingModule({
      providers: [IdGenerator, { provide: APP_ID, useValue: 'ng' }]
    });
    const service = TestBed.inject(IdGenerator);
    const prefix = '';
    const id1 = service.newId(prefix);
    const id2 = service.newId(prefix);
    expect(id1).toBe('-0');
    expect(id2).toBe('-1');
  });

  it('should allow for deterministic snapshots ids 1', async () => {
    const { container } = await render('<test-component/>', {
      imports: [TestComponent]
    });
    expect(container).toMatchSnapshot();
  });
  it('should allow for deterministic snapshots ids 2', async () => {
    const { container } = await render('<test-component/>', {
      imports: [TestComponent]
    });
    expect(container).toMatchSnapshot();
  });
});
