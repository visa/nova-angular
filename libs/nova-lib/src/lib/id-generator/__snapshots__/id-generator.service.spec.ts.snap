// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IdGenerator should allow for deterministic snapshots ids 1 1`] = `
<div
  id="root0"
>
  <test-component>
    <div
      id="a--0"
    />
    <div
      id="a--1"
    />
  </test-component>
</div>
`;

exports[`IdGenerator should allow for deterministic snapshots ids 2 1`] = `
<div
  id="root1"
>
  <test-component>
    <div
      id="a--2"
    />
    <div
      id="a--3"
    />
  </test-component>
</div>
`;
